GMaps Improvements

kvangent opened the pull request:
<!--       PULL REQUESTS CREATED NOT USING ONE OF THE BELOW         --->
<!--      ISSUE TEMPLATES WILL BE CLOSED WITHOUT ANY RESPONSE       --->
<!--       PULL REQUESTS CREATED NOT USING ONE OF THE BELOW         --->
<!--      ISSUE TEMPLATES WILL BE CLOSED WITHOUT ANY RESPONSE       --->
<!--       PULL REQUESTS CREATED NOT USING ONE OF THE BELOW         --->
<!--      ISSUE TEMPLATES WILL BE CLOSED WITHOUT ANY RESPONSE       --->
<!--       PULL REQUESTS CREATED NOT USING ONE OF THE BELOW         --->
<!--      ISSUE TEMPLATES WILL BE CLOSED WITHOUT ANY RESPONSE       --->


<!--
 Please make all PRs to the dev branch. The dev branch will be
 periodically pulled into the master branch. This allows time for
 the changes and documentation to be tested before being exposed to
 a wider population.
--->

## Description
<!-- In detail, describe what your PR adds to PokeAlarm -->
This is a rewrite of the Google Maps API calls used for DTS. It removes the official google python client and instead handles the requests directly to give more control over how they are handled. This will allow for better error handling and won't timeout for extended periods of time when the API Quota has been reached.

Additionally, it removes the automatic DTS detection and instead adds explicit config options to enable the use of DTS. This is to reduce issues users are experiencing with accidentally overusing their quota. (as a side effect I've cleaned up the config.ini.example as well)

Some of the Distance Matrix DTS have been renamed, and are as follows:
* `<walking_distance>` and `<walking_duration>`
* `<biking_distance>` and `<biking_duration>`
* `<driving_distance>` and `<driving_duration>`
* `<transit_distance>` and `<transit_duration>` - (these two are new)

## Type of Change
<!-- Place a single 'x' into the correct box, ex: [x] -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (would cause existing functionality to change)

## Motivation and Context
<!---
 Why is this change required? What problem does it solve?
 If it fixes an open issue, please link to the issue here.
-->
The official GMaps client repeatedly spams the API server until it times out, which can cause several seconds of blocking when trying to trigger a notification. Additionally, some users were hitting their quotas by accident because they were using APIs they didn't know they had enabled.

## Wiki Update
<!--
 Does this feature require an update to the wiki? If so, please submit
 the required change to https://github.com/RocketMap/PokeAlarmWiki.
 If your feature requires a wiki update, you may submit it for review
 but it will not be accepted until the wiki update is complete.
--->
- [X] This change requires an update to the Wiki.
- [ ] This change does not require an update to the Wiki.

The-Real-Ketchum-Dev commented:
The new config file is amazing gonna try this out right now

The-Real-Ketchum-Dev commented:
Works like a charm! Great config.ini and my key isn't spiking anymore

elincognito commented:
I keep getting
`     parse_settings(os.path.abspath(os.path.dirname(__file__)))
   File "start_pokealarm.py", line 287, in parse_settings
     debug=config['DEBUG']
   File "/root/rm2/PokeAlarm/PokeAlarm/Manager.py", line 61, in __init__
     self.set_location(location)
   File "/root/rm2/PokeAlarm/PokeAlarm/Manager.py", line 489, in set_location
     self.__location = self._gmaps_service.get_location_from_name(
 AttributeError: 'GMaps' object has no attribute 'get_location_from_name'`

MARTINI1 commented on config/config.ini.example:
```diff
@@ -1,27 +1,76 @@
+# If a single setting is supplied, it will apply to all Managers.
+# Example: `locale: en` will set all Managers to english
+# If an array of settings is supplied, they will be apply to Managers in order.
+# Example: `locale: [ en, fr, sp ]` sets a different langauge for 3 Managers.
```
`sp` does not exist, I think you want to say spanish so the correct option would be `es`

MARTINI1 commented on config/config.ini.example:
```diff
@@ -1,27 +1,76 @@
+#host: 127.0.0.1                # Interface to listen on (default='127.0.0.1')
+#port: 4000						# Port to listen on (default='4000')
+#concurrency: 200               # Maximum concurrent connections to webserver (default=200)
+#manager_count: 1				# Number of Managers to run (default=1)
```
I think this should go in the manager settings section

kvangent replied:
I left it there because technically it is a web-server setting, instructing it how many Managers to spin up. It also doesn't take an array, and if I moved it any lower it would be under the line that says "Anything beneath this line takes multiple arguments."

MARTINI1 commented on config/config.ini.example:
```diff
@@ -1,27 +1,76 @@
+
+# Location Specific
+#####################
+location:                       # Location, as address or coordinates (default=None)
```
this option not commented with #

elincognito replied:
that explains my error above didn't notice 👀

MARTINI1 commented on config/config.ini.example:
```diff
@@ -1,27 +1,76 @@
-#timezone:                                      # Timezone used for notifications Ex: 'America/Los_Angeles' or '[America/Los_Angeles, America/New_York]'
\ No newline at end of file
+# By default, PA will use `config/config.ini` to load settings.
+# You can override the config file selction with `--config-file` or `-cf`.
```
`selction` -> `selection`

MARTINI1 commented:
Besides the comments I think you have forgotten to close the section titles as webserver or manager

like this:
```
########################
# Webserver Settings
########################
```

The-Real-Ketchum-Dev commented:
It has broke the custom_dts for me

kvangent commented:
Added regular geocoding and surrounded functions with locks to prevent redundant calls from rules, and addressed @MARTINI1's feedback.

@elincognito The error has been cleaned up, so it should be easier to tell what went wrong.

@The-Real-Ketchum-Dev I wasn't able to replicated the issue with custom_dts. Are you sure you have it set up correctly?

The-Real-Ketchum-Dev commented:
@kvangent unless something has changed with custom_dts then no.. It worked before adding this PR

kvangent commented:
@The-Real-Ketchum-Dev Please confirm the same config works on the current `dev` branch, and then switch to this branch and verify it breaks when you have a chance.

The-Real-Ketchum-Dev commented:
Will do.. I'm currently at work but will do so when i get home

On Feb 22, 2018 7:52 PM, "kvangent" <notifications@github.com> wrote:

> @The-Real-Ketchum-Dev <https://github.com/the-real-ketchum-dev> Please
> confirm the same config works on the current dev branch, and then switch
> to this branch and verify it breaks when you have a chance.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PokeAlarm/PokeAlarm/pull/612#issuecomment-367873781>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AimmcyvBw0Yy22eskNAd6N517LL8SzRnks5tXgvdgaJpZM4SLPyO>
> .
>

MARTINI1 commented:
changes added in wiki
